### PR TITLE
filter_record_modifier: allocate bool_map dynamically(#3968)

### LINKED ISSUE #3968: Fluent Bit restarts if a json message has large number of keys
## Bug Report

**Describe the bug**
Fluent Bit restarts if a json message has large number of keys. 

Message: {"message":"DDP connection failed.","timestamp":"2021-08-03T13:33:23.413Z","level":"error","context":{"0":"N","1":"e","2":"t","3":"w","4":"o","5":"r","6":"k","7":" ","8":"e","9":"r","10":"r","11":"o","12":"r","13":":","14":" ","15":"w","16":"s","17":"s","18":":","19":"/","20":"/","21":"d","22":"e","23":"v","24":"s","25":"t","26":"a","27":"b","28":"l","29":"e","30":".","31":"e","32":"d","33":"u","34":".","35":"b","36":"a","37":"r","38":"c","39":"o","40":".","41":"c","42":"l","43":"o","44":"u","45":"d","46":":","47":"4","48":"4","49":"3","50":"/","51":"w","52":"e","53":"b","54":"s","55":"o","56":"c","57":"k","58":"e","59":"t","60":":","61":" ","62":"H","63":"o","64":"s","65":"t","66":"n","67":"a","68":"m","69":"e","70":"/","71":"I","72":"P","73":" ","74":"d","75":"o","76":"e","77":"s","78":" ","79":"n","80":"o","81":"t","82":" ","83":"m","84":"a","85":"t","86":"c","87":"h","88":" ","89":"c","90":"e","91":"r","92":"t","93":"i","94":"f","95":"i","96":"c","97":"a","98":"t","99":"e","100":"'","101":"s","102":" ","103":"a","104":"l","105":"t","106":"n","107":"a","108":"m","109":"e","110":"s","111":":","112":" ","113":"H","114":"o","115":"s","116":"t","117":":","118":" ","119":"d","120":"e","121":"v","122":"s","123":"t","124":"a","125":"b","126":"l","127":"e","128":".","129":"e","130":"d","131":"u","132":".","133":"b","134":"a","135":"r","136":"c","137":"o","138":".","139":"c","140":"l","141":"o","142":"u","143":"d","144":".","145":" ","146":"i","147":"s","148":" ","149":"n","150":"o","151":"t","152":" ","153":"i","154":"n","155":" ","156":"t","157":"h","158":"e","159":" ","160":"c","161":"e","162":"r","163":"t","164":"'","165":"s","166":" ","167":"a","168":"l","169":"t","170":"n","171":"a","172":"m","173":"e","174":"s","175":":","176":" ","177":"D","178":"N","179":"S","180":":","181":"e","182":"d","183":"u","184":".","185":"b","186":"a","187":"r","188":"c","189":"o","190":".","191":"c","192":"o","193":"m","194":",","195":" ","196":"D","197":"N","198":"S","199":":","200":"*","201":".","202":"e","203":"d","204":"u","205":".","206":"b","207":"a","208":"r","209":"c","210":"o","211":".","212":"c","213":"o","214":"m","215":",","216":" ","217":"D","218":"N","219":"S","220":":","221":"*","222":".","223":"c","224":"l","225":"i","226":"c","227":"k","228":"s","229":"h","230":"a","231":"r","232":"e","233":".","234":"a","235":"p","236":"p","237":",","238":" ","239":"D","240":"N","241":"S","242":":","243":"*","244":".","245":"n","246":"e","247":"x","248":"x","249":"i","250":"s","251":".","252":"b","253":"a","254":"r","255":"c","256":"o","257":".","258":"c","259":"o","260":"m","261":",","262":" ","263":"D","264":"N","265":"S","266":":","267":"o","268":"v","269":"e","270":"r","271":"t","272":"u","273":"r","274":"e","275":".","276":"b","277":"a","278":"r","279":"c","280":"o","281":".","282":"c","283":"o","284":"m","285":",","286":" ","287":"D","288":"N","289":"S","290":":","291":"*","292":".","293":"m","294":"e","295":"d","296":"i","297":"a","298":"l","299":"o","300":"n","301":".","302":"b","303":"a","304":"r","305":"c","306":"o","307":".","308":"c","309":"o","310":"m","311":",","312":" ","313":"D","314":"N","315":"S","316":":","317":"s","318":"t","319":"o","320":"r","321":"e","322":".","323":"b","324":"a","325":"r","326":"c","327":"o","328":".","329":"c","330":"o","331":"m","332":",","333":" ","334":"D","335":"N","336":"S","337":":","338":"*","339":".","340":"p","341":"r","342":"o","343":"j","344":".","345":"b","346":"a","347":"r","348":"c","349":"o","350":".","351":"c","352":"o","353":"m","354":",","355":" ","356":"D","357":"N","358":"S","359":":","360":"u","361":"n","362":"i","363":"v","364":"e","365":"r","366":"s","367":"i","368":"t","369":"y","370":".","371":"b","372":"a","373":"r","374":"c","375":"o","376":".","377":"c","378":"o","379":"m","380":",","381":" ","382":"D","383":"N","384":"S","385":":","386":"c","387":"m","388":"p","389":".","390":"b","391":"a","392":"r","393":"c","394":"o","395":".","396":"c","397":"o","398":"m","399":",","400":" ","401":"D","402":"N","403":"S","404":":","405":"i","406":"n","407":"s","408":"i","409":"g","410":"h","411":"t","412":"s","413":".","414":"b","415":"a","416":"r","417":"c","418":"o","419":".","420":"c","421":"o","422":"m","423":",","424":" ","425":"D","426":"N","427":"S","428":":","429":"*","430":".","431":"i","432":"n","433":"s","434":"i","435":"g","436":"h","437":"t","438":"s","439":".","440":"b","441":"a","442":"r","443":"c","444":"o","445":".","446":"c","447":"o","448":"m","449":",","450":" ","451":"D","452":"N","453":"S","454":":","455":"*","456":".","457":"w","458":"s","459":"g","460":".","461":"b","462":"a","463":"r","464":"c","465":"o","466":".","467":"c","468":"o","469":"m","470":",","471":" ","472":"D","473":"N","474":"S","475":":","476":"m","477":"e","478":"d","479":"i","480":"a","481":"l","482":"o","483":"n","484":".","485":"b","486":"a","487":"r","488":"c","489":"o","490":".","491":"c","492":"o","493":"m","494":",","495":" ","496":"D","497":"N","498":"S","499":":","500":"n","501":"e","502":"x","503":"x","504":"i","505":"s","506":".","507":"b","508":"a","509":"r","510":"c","511":"o","512":".","513":"c","514":"o","515":"m","516":",","517":" ","518":"D","519":"N","520":"S","521":":","522":"a","523":"c","524":"t","525":"i","526":"v","527":"a","528":"t","529":"e","530":".","531":"b","532":"a","533":"r","534":"c","535":"o","536":".","537":"c","538":"o","539":"m","540":",","541":" ","542":"D","543":"N","544":"S","545":":","546":"*","547":".","548":"c","549":"l","550":"o","551":"u","552":"d","553":".","554":"b","555":"a","556":"r","557":"c","558":"o","559":".","560":"c","561":"o","562":"m","563":",","564":" ","565":"D","566":"N","567":"S","568":":","569":"*","570":".","571":"o","572":"v","573":"e","574":"r","575":"t","576":"u","577":"r","578":"e","579":".","580":"b","581":"a","582":"r","583":"c","584":"o","585":".","586":"c","587":"o","588":"m","589":",","590":" ","591":"D","592":"N","593":"S","594":":","595":"c","596":"l","597":"i","598":"c","599":"k","600":"s","601":"h","602":"a","603":"r","604":"e","605":".","606":"a","607":"p","608":"p","609":",","610":" ","611":"D","612":"N","613":"S","614":":","615":"*","616":".","617":"u","618":"n","619":"i","620":"v","621":"e","622":"r","623":"s","624":"i","625":"t","626":"y","627":".","628":"b","629":"a","630":"r","631":"c","632":"o","633":".","634":"c","635":"o","636":"m","637":",","638":" ","639":"D","640":"N","641":"S","642":":","643":"*","644":".","645":"s","646":"t","647":"o","648":"r","649":"e","650":".","651":"b","652":"a","653":"r","654":"c","655":"o","656":".","657":"c","658":"o","659":"m","660":",","661":" ","662":"D","663":"N","664":"S","665":":","666":"x","667":"m","668":"s","669":".","670":"b","671":"a","672":"r","673":"c","674":"o","675":".","676":"c","677":"o","678":"m","679":",","680":" ","681":"D","682":"N","683":"S","684":":","685":"*","686":".","687":"a","688":"c","689":"t","690":"i","691":"v","692":"a","693":"t","694":"e","695":".","696":"b","697":"a","698":"r","699":"c","700":"o","701":".","702":"c","703":"o","704":"m","705":",","706":" ","707":"D","708":"N","709":"S","710":":","711":"w","712":"e","713":"c","714":"o","715":"n","716":"n","717":"e","718":"c","719":"t","720":".","721":"b","722":"a","723":"r","724":"c","725":"o","726":".","727":"c","728":"o","729":"m","730":",","731":" ","732":"D","733":"N","734":"S","735":":","736":"*","737":".","738":"w","739":"e","740":"c","741":"o","742":"n","743":"n","744":"e","745":"c","746":"t","747":".","748":"b","749":"a","750":"r","751":"c","752":"o","753":".","754":"c","755":"o","756":"m","757":",","758":" ","759":"D","760":"N","761":"S","762":":","763":"*","764":".","765":"c","766":"m","767":"p","768":".","769":"b","770":"a","771":"r","772":"c","773":"o","774":".","775":"c","776":"o","777":"m","778":",","779":" ","780":"D","781":"N","782":"S","783":":","784":"w","785":"s","786":"g","787":".","788":"b","789":"a","790":"r","791":"c","792":"o","793":".","794":"c","795":"o","796":"m","797":",","798":" ","799":"D","800":"N","801":"S","802":":","803":"p","804":"r","805":"o","806":"j","807":".","808":"b","809":"a","810":"r","811":"c","812":"o","813":".","814":"c","815":"o","816":"m","817":",","818":" ","819":"D","820":"N","821":"S","822":":","823":"c","824":"l","825":"o","826":"u","827":"d","828":".","829":"b","830":"a","831":"r","832":"c","833":"o","834":".","835":"c","836":"o","837":"m","838":",","839":" ","840":"D","841":"N","842":"S","843":":","844":"*","845":".","846":"x","847":"m","848":"s","849":".","850":"b","851":"a","852":"r","853":"c","854":"o","855":".","856":"c","857":"o","858":"m","859":",","860":" ","861":"D","862":"N","863":"S","864":":","865":"*","866":".","867":"b","868":"i","869":"n","870":".","871":"c","872":"l","873":"o","874":"u","875":"d","876":".","877":"b","878":"a","879":"r","880":"c","881":"o","882":".","883":"c","884":"o","885":"m","service":"ROOT","id":"vwhjvhvjs","serviceMetadata":{"meetingId":"","instituteId":"obyr","role":"abc"}}}

**To Reproduce**
- Add a input tail plugin that reads from a file
- Paste the above message in that file
- Check Fluent Bit Status
```
Aug 18 11:51:33 nrc-000babcd6ada barco-nrc-td-agent-bit[13826]: [2021/08/18 11:51:33] [debug] [storage] [cio file] alloc_size from 4096 to 36864
Aug 18 11:51:33 nrc-000babcd6ada barco-nrc-td-agent-bit[13826]: [2021/08/18 11:51:33] [debug] [filter:nest:nest.5] Lift : Outer map size is 5, will be 893, lifting 1 record(s)
Aug 18 11:51:33 nrc-000babcd6ada barco-nrc-td-agent-bit[13826]: [2021/08/18 11:51:33] [debug] [filter:nest:nest.6] Lift : Outer map size is 893, will be 895, lifting 1 record(s)
Aug 18 11:51:33 nrc-000babcd6ada barco-nrc-td-agent-bit[13826]: [2021/08/18 11:51:33] [engine] caught signal (SIGSEGV)
Aug 18 11:51:33 nrc-000babcd6ada barco-nrc-td-agent-bit[13826]: #0  0x562822076522      in  cb_modifier_filter() at plugins/filter_record_modifier/filter_modifier.c:307
Aug 18 11:51:33 nrc-000babcd6ada systemd[1]: barco-nrc-td-agent-bit.service: Main process exited, code=killed, status=11/SEGV
Aug 18 11:51:33 nrc-000babcd6ada systemd[1]: barco-nrc-td-agent-bit.service: Failed with result 'signal'.

```
- Steps to reproduce the problem:

**Expected behavior**
Fluent Bit should gracefully handle when using "nest" filter and "Nested_under" key has large number of keys 

**Screenshots**

**Your Environment**

* Version used: 1.8.1, 1.8.3
* Configuration: 
* [INPUT]
    name tail
    storage.type filesystem
    Mem_Buf_Limit 5M
    Buffer_Chunk_Size 32KB
    Buffer_Max_Size 1M
    Path /var/log/abc/def*.log
    DB /etc/barco-nrc-td-agent-bit/log-abc.db
    Skip_Long_Lines on
    Path_Key file
    tag  app-logs
    Parser app_parser

[FILTER]
    Name nest
    Match app-logs
    Operation lift
    Nested_under context

[FILTER]
    Name nest
    Match app-logs
    Operation lift
    Nested_under serviceMetadata

[FILTER]
    Name record_modifier
    Match app-logs
    Record version ${VERSION}
    Record env ${ENV}
    Record component xyz

[FILTER]
    Name nest
    Match app-logs
    Operation nest
    Wildcard version
    Wildcard namw
    Nest_under serviceMetadata

[FILTER]
    Name nest
    Match app-logs
    Operation nest
    Wildcard serviceMetadata
    Nest_under context

* Environment name and version (e.g. Kubernetes? What version?): Tail

* Server type and version:
* Operating System and version:
* Filters and plugins:

**Additional context**
I tried using "grep" plugin to discard any such message.
[FILTER]
    Name grep
    Match *
    Exclude context ([^:]*:){125}([^:]*)  
Aim is to discard any message that has more than 125 keys but it is not working. I am not sure if regular expression in the Exclude statement is not working or my approach is incorrect.
Also, after doing some google search, I changed it to Exclude context \([^:]*:){125}([^:]*)\ but even that didn't work.
  
What are the best practices to handle these scenarios?
Is there any limit on number of keys that fluent bit can handle?

### PR DESCRIPTION
**Pull request by nokute78**
Fixes #3968 

Currently, the size of `bool_map`  is fixed.
If the map size of incoming record exceeds `bool_map` size,
it breaks stack.

This patch is to allocate `bool_map` dynamically.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [X] Example configuration file for the change
- [X] Debug log output from testing the change

- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [N/A] Documentation required for this feature

## Configuration

https://github.com/fluent/fluent-bit/issues/3968#issuecomment-903007287

## Debug log

Not SIGABRT.

```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/08/21 10:20:25] [ info] [engine] started (pid=73234)
[2021/08/21 10:20:25] [ info] [storage] version=1.1.1, initializing...
[2021/08/21 10:20:25] [ info] [storage] in-memory
[2021/08/21 10:20:25] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/08/21 10:20:25] [ info] [cmetrics] version=0.2.1
[2021/08/21 10:20:25] [ info] [input:http:http.0] listening on 0.0.0.0:9880
[2021/08/21 10:20:25] [ info] [sp] stream processor started
[0] json.log: [1629508830.376201736, {"message"=>"DDP connection failed.", "timestamp"=>"2021-08-03T13:33:23.413Z", "level"=>"error", "0"=>"N", "1"=>"e", "2"=>"t", "3"=>"w", "4"=>"o", "5"=>"r", "6"=>"k", "7"=>" ", "8"=>"e", "9"=>"r", "10"=>"r", "11"=>"o", "12"=>"r", "13"=>":", "14"=>" ", "15"=>"w", "16"=>"s", "17"=>"s", "18"=>":", "19"=>"/", "20"=>"/", "21"=>"d", "22"=>"e", "23"=>"v", "24"=>"s", "25"=>"t", "26"=>"a", "27"=>"b", "28"=>"l", "29"=>"e", "30"=>".", "31"=>"e", "32"=>"d", "33"=>"u", "34"=>".", "35"=>"b", "36"=>"a", "37"=>"r", "38"=>"c", "39"=>"o", "40"=>".", "41"=>"c", "42"=>"l", "43"=>"o", "44"=>"u", "45"=>"d", "46"=>":", "47"=>"4", "48"=>"4", "49"=>"3", "50"=>"/", "51"=>"w", "52"=>"e", "53"=>"b", "54"=>"s", "55"=>"o", "56"=>"c", "57"=>"k", "58"=>"e", "59"=>"t", "60"=>":", "61"=>" ", "62"=>"H", "63"=>"o", "64"=>"s", "65"=>"t", "66"=>"n", "67"=>"a", "68"
(snip)
```

## Valgrind output

No leak.

```
$ valgrind --leak-check=full ../bin/fluent-bit -c a.conf 
==74414== Memcheck, a memory error detector
==74414== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==74414== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==74414== Command: ../bin/fluent-bit -c a.conf
==74414== 
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/08/21 10:23:49] [ info] [engine] started (pid=74414)
[2021/08/21 10:23:50] [ info] [storage] version=1.1.1, initializing...
[2021/08/21 10:23:50] [ info] [storage] in-memory
[2021/08/21 10:23:50] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/08/21 10:23:50] [ info] [cmetrics] version=0.2.1
[2021/08/21 10:23:50] [ info] [input:http:http.0] listening on 0.0.0.0:9880
[2021/08/21 10:23:50] [ info] [sp] stream processor started
==74414== Warning: client switching stacks?  SP change: 0x57e49d8 --> 0x4e37070
==74414==          to suppress, use: --max-stackframe=10148200 or greater
==74414== Warning: client switching stacks?  SP change: 0x4e36fe8 --> 0x57e49d8
==74414==          to suppress, use: --max-stackframe=10148336 or greater
==74414== Warning: client switching stacks?  SP change: 0x57e49d8 --> 0x4e36fe8
==74414==          to suppress, use: --max-stackframe=10148336 or greater
==74414==          further instances of this message will not be shown.
[0] json.log: [1629509038.978531832, {"message"=>"DDP connection failed.", "timestamp"=>"2021-08-03T13:33:23.413Z", "level"=>"error", "0"=>"N", "1"=>"e", "2"=>"t", "3"=>"w", "4"=>"o", "5"=>"r", "6"=>"k", "7"=>" ", "8"=>"e", "9"=>"r", "10"=>"r", "11"=>"o", "12"=>"r", "13"=>":", "14"=>" ", "15"=>"w", "16"=>"s", "17"=>"s", "18"=>":", "19"=>"/", "20"=>"/", "21"=>"d", "22"=>"e", "23"=>"v", "24"=>"s", "25"=>"t", "26"=>"a", "27"=>"b", "28"=>"l", "29"=>"e", "30"=>".", "
(snip)
"824"=>"l", "825"=>"o", "826"=>"u", "827"=>"d", "828"=>".", "829"=>"b", "830"=>"a", "831"=>"r", "832"=>"c", "833"=>"o", "834"=>".", "835"=>"c", "836"=>"o", "837"=>"m", "838"=>",", "839"=>" ", "840"=>"D", "841"=>"N", "842"=>"S", "843"=>":", "844"=>"", "845"=>".", "846"=>"x", "847"=>"m", "848"=>"s", "849"=>".", "850"=>"b", "851"=>"a", "852"=>"r", "853"=>"c", "854"=>"o", "855"=>".", "856"=>"c", "857"=>"o", "858"=>"m", "859"=>",", "860"=>" ", "861"=>"D", "862"=>"N", "863"=>"S", "864"=>":", "865"=>"", "866"=>".", "867"=>"b", "868"=>"i", "869"=>"n", "870"=>".", "871"=>"c", "872"=>"l", "873"=>"o", "874"=>"u", "875"=>"d", "876"=>".", "877"=>"b", "878"=>"a", "879"=>"r", "880"=>"c", "881"=>"o", "882"=>".", "883"=>"c", "884"=>"o", "885"=>"m", "service"=>"ROOT", "id"=>"vwhjvhvjs", "serviceMetadata"=>{"meetingId"=>"", "instituteId"=>"obyr", "role"=>"abc"}}]
^C[2021/08/21 10:24:10] [engine] caught signal (SIGINT)
[2021/08/21 10:24:10] [ warn] [engine] service will stop in 5 seconds
[2021/08/21 10:24:14] [ info] [engine] service stopped
==74414== 
==74414== HEAP SUMMARY:
==74414==     in use at exit: 0 bytes in 0 blocks
==74414==   total heap usage: 1,286 allocs, 1,286 frees, 3,005,936 bytes allocated
==74414== 
==74414== All heap blocks were freed -- no leaks are possible
==74414== 
==74414== For lists of detected and suppressed errors, rerun with: -s
==74414== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: plugins/filter_record_modifier/filter_modifier.c [hunk 1]
```diff
@@ -293,6 +298,12 @@ static int cb_modifier_filter(const void *data, size_t bytes,
         /* grep keys */
         if (obj->type == MSGPACK_OBJECT_MAP) {
             map_num = obj->via.map.size;
+            /* allocate map_num + guard byte */
+            bool_map = flb_calloc(map_num+1, sizeof(bool_map_t));
```
Comment (edsiper) [inline plugins/filter_record_modifier/filter_modifier.c:302]: I think we should have a limit, what do you think ?

Comment (nokute78) [reply plugins/filter_record_modifier/filter_modifier.c:302]: I agree.

Do you have any idea about amount of limit ?
Is 4096 enough ? 65536 ? or configurable ? e.g. `map_size_limit 4096`

Comment (nokute78) [reply plugins/filter_record_modifier/filter_modifier.c:308]: I updated PR to add fixed limit 65535.

**Comment by ujjwalgarg1995** [comment]
@edsiper @nokute78 any update on this?

**Comment by nokute78** [comment]
#4013 and https://github.com/aws/aws-for-fluent-bit/issues/224 may be similar issue.

**Comment by ujjwalgarg1995** [comment]
@edsiper  @JeffLuoo  can one you please review this?

**Comment by nokute78** [comment]
#3554 may be related.
@edsiper Could you review it ? So many issues about this had been reported..

**Comment by edsiper** [comment]
thanks

**Pull request merged**